Calico SIGSEV on null address in Kubernetes (3.30.1)

Line of code with the SIGSEV is here - https://github.com/projectcalico/calico/blob/master/felix/calc/l3_route_resolver.go#L166
 
### calico-node unhealthy across the cluster
```
Events:
Type Reason Age From Message
 
Warning Unhealthy 90s (x604 over 4h8m) kubelet (combined from similar events): Readiness probe failed: 2025-09-17 04:27:39.099 [INFO][3135853] node/health.go 202: Number of node(s) with BGP peering established = 58
calico/node is not ready: felix is not ready: readiness probe reporting 503
```
 
### Stack Trace
```
2025-09-17 04:51:26.668 [INFO][3309420] felix/int_dataplane.go 2201: Received *proto.ServiceUpdate update from calculation graph. msg=name:"kubernetes-dashboard-kong-proxy" namespace:"kubernetes-dashboard" type:"ClusterIP" cluster_ips:"10.150.245.165" ports:{Protocol:"TCP" Port:443}
2025-09-17 04:51:26.668 [INFO][3309420] felix/int_dataplane.go 2201: Received *proto.ServiceUpdate update from calculation graph. msg=name:"kube-prometheus-stack-operator" namespace:"prometheus" type:"ClusterIP" cluster_ips:"10.150.180.202" ports:{Protocol:"TCP" Port:443}
2025-09-17 04:51:26.668 [INFO][3309420] felix/int_dataplane.go 2201: Received *proto.ServiceUpdate update from calculation graph. msg=name:"runai-backend-workloads" namespace:"runai-backend" type:"ClusterIP" cluster_ips:"10.150.147.134" ports:{Protocol:"TCP" Port:8080} ports:{Protocol:"TCP" Port:9090}
2025-09-17 04:51:26.668 [INFO][3309420] felix/int_dataplane.go 2201: Received *proto.ServiceUpdate update from calculation graph. msg=name:"runai-job-controller" namespace:"runai" type:"ClusterIP" cluster_ips:"10.150.135.80" ports:{Protocol:"TCP" Port:9797}
2025-09-17 04:51:26.680 [INFO][3309420] felix/vxlan_resolver.go 304: Missing both IPv4 and IPv6 VTEP information for node, cannot send VTEP yet node="kr1bk8s00dgx001"
2025-09-17 04:51:26.680 [INFO][3309420] felix/vxlan_resolver.go 304: Missing both IPv4 and IPv6 VTEP information for node, cannot send VTEP yet node="kr1bk8s00dgx002"
2025-09-17 04:51:26.680 [INFO][3309420] felix/vxlan_resolver.go 304: Missing both IPv4 and IPv6 VTEP information for node, cannot send VTEP yet node="kr1bk8s00dgx003"
2025-09-17 04:51:26.680 [INFO][3309420] felix/vxlan_resolver.go 304: Missing both IPv4 and IPv6 VTEP information for node, cannot send VTEP yet node="kr1bk8s00dgx004"
2025-09-17 04:51:26.680 [INFO][3309420] felix/vxlan_resolver.go 304: Missing both IPv4 and IPv6 VTEP information for node, cannot send VTEP yet node="kr1bk8s00dgx005"
2025-09-17 04:51:26.680 [INFO][3309420] felix/vxlan_resolver.go 304: Missing both IPv4 and IPv6 VTEP information for node, cannot send VTEP yet node="kr1bk8s00dgx006"
2025-09-17 04:51:26.680 [INFO][3309420] felix/vxlan_resolver.go 304: Missing both IPv4 and IPv6 VTEP information for node, cannot send VTEP yet node="kr1bk8s00dgx007"
2025-09-17 04:51:26.680 [INFO][3309420] felix/vxlan_resolver.go 304: Missing both IPv4 and IPv6 VTEP information for node, cannot send VTEP yet node="kr1bk8s00dgx008"
2025-09-17 04:51:26.681 [INFO][3309420] felix/vxlan_resolver.go 304: Missing both IPv4 and IPv6 VTEP information for node, cannot send VTEP yet node="kr1bk8s00dgx009"
2025-09-17 04:51:26.681 [INFO][3309420] felix/vxlan_resolver.go 304: Missing both IPv4 and IPv6 VTEP information for node, cannot send VTEP yet node="kr1bk8s00dgx010"
2025-09-17 04:51:26.681 [INFO][3309420] felix/vxlan_resolver.go 304: Missing both IPv4 and IPv6 VTEP information for node, cannot send VTEP yet node="kr1bk8s00dgx011"
2025-09-17 04:51:26.681 [INFO][3309420] felix/vxlan_resolver.go 304: Missing both IPv4 and IPv6 VTEP information for node, cannot send VTEP yet node="kr1bk8s00dgx012"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x20d0ca9]
 
goroutine 369 [running]:
[github.com/projectcalico/calico/felix/calc.l3rrNodeInfo.AddressesAsCIDRs](http://github.com/projectcalico/calico/felix/calc.l3rrNodeInfo.AddressesAsCIDRs)({{0xa, 0xa, 0x20, 0x18}, {{0xa, 0xa, 0x20, 0x0}, 0x1a}, {0x0, ...}, ...})
/go/src/[github.com/projectcalico/calico/felix/calc/l3_route_resolver.go:166](http://github.com/projectcalico/calico/felix/calc/l3_route_resolver.go:166) +0x309
[github.com/projectcalico/calico/felix/calc](http://github.com/projectcalico/calico/felix/calc).(*L3RouteResolver).onNodeUpdate(0xc000db4120, {0xc000a1e9ad, 0xf}, 0xc0010cd978)
/go/src/[github.com/projectcalico/calico/felix/calc/l3_route_resolver.go:578](http://github.com/projectcalico/calico/felix/calc/l3_route_resolver.go:578) +0xe25
[github.com/projectcalico/calico/felix/calc](http://github.com/projectcalico/calico/felix/calc).(*L3RouteResolver).OnResourceUpdate(0xc000db4120, {{{0x509faa0, 0xc001d98bd0}, {0x488d780, 0xc001601180}, {0xc000dff980, 0x9}, 0x0, 0x0}, 0x1})
/go/src/[github.com/projectcalico/calico/felix/calc/l3_route_resolver.go:442](http://github.com/projectcalico/calico/felix/calc/l3_route_resolver.go:442) +0xae5
[github.com/projectcalico/calico/felix/dispatcher.updateHandlers.DispatchToAll](http://github.com/projectcalico/calico/felix/dispatcher.updateHandlers.DispatchToAll)(...)
/go/src/[github.com/projectcalico/calico/felix/dispatcher/dispatcher.go:45](http://github.com/projectcalico/calico/felix/dispatcher/dispatcher.go:45)
[github.com/projectcalico/calico/felix/dispatcher](http://github.com/projectcalico/calico/felix/dispatcher).(*Dispatcher).OnUpdate(0xc0002500c0, {{{0x509faa0, 0xc001d98bd0}, {0x488d780, 0xc001601180}, {0xc000dff980, 0x9}, 0x0, 0x0}, 0x1})
/go/src/[github.com/projectcalico/calico/felix/dispatcher/dispatcher.go:105](http://github.com/projectcalico/calico/felix/dispatcher/dispatcher.go:105) +0x395
[github.com/projectcalico/calico/felix/dispatcher](http://github.com/projectcalico/calico/felix/dispatcher).(*Dispatcher).OnUpdates(...)
/go/src/[github.com/projectcalico/calico/felix/dispatcher/dispatcher.go:80](http://github.com/projectcalico/calico/felix/dispatcher/dispatcher.go:80)
[github.com/projectcalico/calico/felix/calc](http://github.com/projectcalico/calico/felix/calc).(*CalcGraph).OnUpdates(...)
/go/src/[github.com/projectcalico/calico/felix/calc/calc_graph.go:137](http://github.com/projectcalico/calico/felix/calc/calc_graph.go:137)
[github.com/projectcalico/calico/felix/calc](http://github.com/projectcalico/calico/felix/calc).(*AsyncCalcGraph).loop(0xc0003fe9a0)
/go/src/[github.com/projectcalico/calico/felix/calc/async_calc_graph.go:151](http://github.com/projectcalico/calico/felix/calc/async_calc_graph.go:151) +0x75c
created by [github.com/projectcalico/calico/felix/calc](http://github.com/projectcalico/calico/felix/calc).(*AsyncCalcGraph).Start in goroutine 1
/go/src/[github.com/projectcalico/calico/felix/calc/async_calc_graph.go:259](http://github.com/projectcalico/calico/felix/calc/async_calc_graph.go:259) +0xf8
```
 
RCA from Internal Investigation  was that the External IP field was set to this accidentally
```yaml
- type: ExternalIP
   address: ""
```
 
Cleaning this value up by removing the `ExternalIP` array entry was the fix.

--- a/felix/calc/l3_route_resolver.go
+++ b/felix/calc/l3_route_resolver.go
@@ -153,9 +153,9 @@ func (i l3rrNodeInfo) AddressesAsCIDRs() []ip.CIDR {
 		addrs[a] = struct{}{}
 	}
 
-	// Clean up empty (uninitialized) addresses
+	// Clean up empty (uninitialized) or nil addresses
 	for a := range addrs {
-		if a == emptyV4Addr || a == emptyV6Addr {
+		if a == nil || a == emptyV4Addr || a == emptyV6Addr {
 			delete(addrs, a)
 		}
 	}

--- a/felix/calc/l3_route_resolver_test.go
+++ b/felix/calc/l3_route_resolver_test.go
@@ -154,6 +154,17 @@ var _ = Describe("L3RouteResolver", func() {
 			}
 			Expect(info.AddressesAsCIDRs()).To(Equal([]ip.CIDR{}))
 		})
+		It("should not panic on nil addresses in AddressesAsCIDRs()", func() {
+			// Test for issue #11384: SIGSEGV on null address
+			// When ExternalIP is empty, a nil ip.Addr can end up in the Addresses slice.
+			// The nil check should prevent a panic when AsCIDR() is called.
+			info := l3rrNodeInfo{
+				Addresses: []ip.Addr{nil, ip.FromString("192.168.0.1"), nil},
+			}
+			cidrs := info.AddressesAsCIDRs()
+			Expect(cidrs).To(HaveLen(1))
+			Expect(cidrs[0].String()).To(Equal("192.168.0.1/32"))
+		})
 		It("should consider VXLANV6Addr in Equal() method", func() {
 			info1 := l3rrNodeInfo{
 				VXLANV6Addr: ip.FromString("dead:beef::1"),

